File: src/main/java/com/bluehabit/budgetku/component/user/service/UserService.java
/*
 * Copyright © 2023 Blue Habit.
 *
 * Unauthorized copying, publishing of this file, via any medium is strictly prohibited
 * Proprietary and confidential
 */

package com.bluehabit.budgetku.component.user.service;

import com.bluehabit.budgetku.common.BaseResponse;
import com.bluehabit.budgetku.common.GoogleAuthUtil;
import com.bluehabit.budgetku.component.user.model.*;
import com.bluehabit.budgetku.config.JwtService;
import com.bluehabit.budgetku.component.user.entity.UserCredential;
import com.bluehabit.budgetku.component.user.entity.UserProfile;
import com.bluehabit.budgetku.exception.UnAuthorizedException;
import com.bluehabit.budgetku.component.user.repo.UserCredentialRepository;
import com.bluehabit.budgetku.component.user.repo.UserProfileRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Pageable;
import org.springframework.http.ResponseEntity;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Service;

import java.time.OffsetDateTime;
import java.util.List;
import java.util.UUID;

@Service
public class UserService {
    @Autowired
    private UserCredentialRepository userCredentialRepository;
    @Autowired
    private UserProfileRepository userProfileRepository;
    @Autowired
    private BCryptPasswordEncoder encoder;
    @Autowired
    private JwtService jwtService;


    public ResponseEntity<BaseResponse<SignInResponse>> signInWithEmail(SignInWithEmailRequest request) {
        var user = userCredentialRepository.findByUserEmail(request.email());

        if (user.isEmpty()) {
            throw new UnAuthorizedException("Sign in failed, user not found!");
        }
        if (!encoder.matches(request.password(), user.get().getUserPassword())) {
            throw new UnAuthorizedException("Email or Password invalid");
        }
        if (!user.get().getUserAuthProvider().equals("BASIC")) {
            throw new UnAuthorizedException("Email already registered to another account");
        }

        var token = jwtService.generateToken(user.get().getUserEmail());
        return BaseResponse.success("Sign in with email success", new SignInResponse(token, user.get()));
    }

    public ResponseEntity<BaseResponse<SignInResponse>> signInWithGoogle(SignInWithGoogleRequest request) {
        var googleAuthClaim = GoogleAuthUtil.getGoogleClaim(request.token());

        if (googleAuthClaim.isEmpty()) {
            throw new UnAuthorizedException("Given token is invalid");
        }

        var findUser = userCredentialRepository.findByUserEmail(googleAuthClaim.get().email());
        if (findUser.isEmpty()) {
            throw new UnAuthorizedException("User not registered!");
        }

        if (!findUser.get().getUserAuthProvider().equals("GOOGLE")) {
            throw new UnAuthorizedException("Email already registered to another account");
        }

        var token = jwtService.generateToken(findUser.get().getUserEmail());
        return BaseResponse.success("Sign in with email success", new SignInResponse(token, findUser.get()));
    }

    public ResponseEntity<BaseResponse<UserCredential>> signUpWithGoogle(SignUpWithGoogleRequest req) {
        var claims = GoogleAuthUtil.getGoogleClaim(req.token());
        if (claims.isEmpty()) {
            throw  new UnAuthorizedException(1, "Token is invalid");
        }

        var findUser = userCredentialRepository.exist(claims.get().email());
        if (findUser) {
            throw  new UnAuthorizedException(2, "Already registered");
        }
        var uuid = UUID.randomUUID().toString();
        var currentDate = OffsetDateTime.now();

        var userProfile = new UserProfile();
        userProfile.setUserId(uuid);
        userProfile.setUserFullName(claims.get().fullName());
        userProfile.setUserProfilePicture(null);
        userProfile.setUserDateOfBirth(null);
        userProfile.setUserCountryCode(claims.get().locale());
        userProfile.setUserPhoneUmber(null);
        userProfile.setCreatedAt(currentDate);
        userProfile.setUpdatedAt(currentDate);

        var savedProfile = userProfileRepository.save(userProfile);
        var userCredential = new UserCredential();
        userCredential.setUserId(uuid);
        userCredential.setUserEmail(claims.get().email());
        userCredential.setUserPassword(encoder.encode(claims.get().email()));
        userCredential.setUserStatus("ACTIVE");
        userCredential.setUserAuthProvider("GOOGLE");
        userCredential.setUserProfile(savedProfile);
        userCredential.setUserNotificationToken("");
        userCredential.setCreatedAt(currentDate);
        userCredential.setUpdatedAt(currentDate);
        var savedCredential = userCredentialRepository.save(userCredential);


        return BaseResponse.success("Success", savedCredential);
    }

    public ResponseEntity<BaseResponse<UserCredential>> signUpWithEmail(SignUpWithEmailRequest req) {

        var findUser = userCredentialRepository.exist(req.email());
        if (findUser) {
            throw new UnAuthorizedException(2, "Already registered");
        }
        var uuid = UUID.randomUUID().toString();
        var currentDate = OffsetDateTime.now();

        var userProfile = new UserProfile();
        userProfile.setUserId(uuid);
        userProfile.setUserFullName(req.fullName());
        userProfile.setUserProfilePicture(null);
        userProfile.setUserDateOfBirth(null);
        userProfile.setUserCountryCode("ID");
        userProfile.setUserPhoneUmber(null);
        userProfile.setCreatedAt(currentDate);
        userProfile.setUpdatedAt(currentDate);

        var savedProfile = userProfileRepository.save(userProfile);
        var userCredential = new UserCredential();
        userCredential.setUserId(uuid);
        userCredential.setUserEmail(req.email());
        userCredential.setUserPassword(encoder.encode(req.password()));
        userCredential.setUserStatus("ACTIVE");
        userCredential.setUserAuthProvider("GOOGLE");
        userCredential.setUserProfile(savedProfile);
        userCredential.setUserNotificationToken("");
        userCredential.setCreatedAt(currentDate);
        userCredential.setUpdatedAt(currentDate);
        var savedCredential = userCredentialRepository.save(userCredential);


        return BaseResponse.success("Success", savedCredential);
    }

    public ResponseEntity<BaseResponse<List<UserProfile>>> getUsers(Pageable pageable) {
        var user = userProfileRepository.findAll(pageable);
        return BaseResponse.success("Get all users", user.toList());
    }

    public ResponseEntity<BaseResponse<String>> refreshToken(String token){
        try{
            if (token == null || token.isEmpty()) {
                throw  new UnAuthorizedException(401,"Token not provided");
            }

            if (!token.startsWith("Bearer")) {
                throw  new UnAuthorizedException(401,"header doesn't contain Bearer");
            }

            String split = token.substring(7);
            String username = jwtService.extractFromExpired(split);

            if (username.isEmpty()) {
                throw  new UnAuthorizedException(401,"failed extract claim");
            }
            var findUser = userCredentialRepository.findByUserEmail(username);

            if(findUser.isEmpty()){
                throw  new UnAuthorizedException(4000,"there is no user with "+username);
            }

            var generatedToken = jwtService.generateToken(findUser.get().getUserEmail());

            return BaseResponse.success("Success",generatedToken);
        }catch (Exception e){
            throw  new UnAuthorizedException(401,e.getClass().getSimpleName());
        }
    }

}

File: src/main/kotlin/com/bluehabit/budgetku/data/user/UserService.kt
/*
 * Copyright © 2023 Blue Habit.
 *
 * Unauthorized copying, publishing of this file, via any medium is strictly prohibited
 * Proprietary and confidential
 */

package com.bluehabit.budgetku.data.user

import com.bluehabit.budgetku.common.Constants.Permission.READ_USER
import com.bluehabit.budgetku.common.exception.UnAuthorizedException
import com.bluehabit.budgetku.common.model.AuthBaseResponse
import com.bluehabit.budgetku.common.model.BaseResponse
import com.bluehabit.budgetku.common.model.PagingDataResponse
import com.bluehabit.budgetku.common.model.baseAuthResponse
import com.bluehabit.budgetku.common.model.baseResponse
import com.bluehabit.budgetku.common.utils.GoogleAuthUtil
import com.bluehabit.budgetku.common.utils.ValidationUtil
import com.bluehabit.budgetku.common.utils.allowTo
import com.bluehabit.budgetku.common.utils.createFileName
import com.bluehabit.budgetku.common.utils.getMimeTypes
import com.bluehabit.budgetku.common.utils.getTodayDateTime
import com.bluehabit.budgetku.common.utils.getTodayDateTimeOffset
import com.bluehabit.budgetku.config.JwtUtil
import com.bluehabit.budgetku.data.BaseService
import com.bluehabit.budgetku.data.user.UserAuthProvider.BASIC
import com.bluehabit.budgetku.data.user.UserStatus.ACTIVE
import com.bluehabit.budgetku.data.user.UserStatus.WAITING_CONFIRMATION
import com.bluehabit.budgetku.data.user.userActivity.UserActivityRepository
import com.bluehabit.budgetku.data.user.userCredential.UserCredential
import com.bluehabit.budgetku.data.user.userCredential.UserCredentialRepository
import com.bluehabit.budgetku.data.user.userProfile.UserProfile
import com.bluehabit.budgetku.data.user.userProfile.UserProfileRepository
import com.bluehabit.budgetku.data.user.userVerification.UserVerification
import com.bluehabit.budgetku.data.user.userVerification.UserVerificationRepository
import jakarta.transaction.Transactional
import org.springframework.context.support.ResourceBundleMessageSource
import org.springframework.core.env.Environment
import org.springframework.data.domain.Pageable
import org.springframework.data.repository.findByIdOrNull
import org.springframework.http.HttpStatus.OK
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.User
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.crypto.scrypt.SCryptPasswordEncoder
import org.springframework.stereotype.Service
import java.time.OffsetDateTime
import java.time.Period
import java.util.*


@Service
class UserService(
    private val userCredentialRepository: UserCredentialRepository,
    private val userActivityRepository: UserActivityRepository,
    private val userVerificationRepository: UserVerificationRepository,
    private val userProfileRepository: UserProfileRepository,
    private val validationUtil: ValidationUtil,
    private val environment: Environment,
    private val i18n: ResourceBundleMessageSource,
    private val scrypt: SCryptPasswordEncoder,
    private val jwtUtil: JwtUtil
) : UserDetailsService, BaseService(
    userCredentialRepository,
    i18n
) {

    //region admin
    @Transactional
    override fun loadUserByUsername(username: String): UserDetails? {

        val user = userCredentialRepository
            .findByUserEmail(username) ?: return null
        return User(
            username,
            user.userPassword,
            user.userStatus == ACTIVE.name,
            false,
            false,
            false,
            user.userPermissions.map { SimpleGrantedAuthority(it.permissionType) }
        )

    }
    //end region

    //region user auth
    fun signInWithEmail(
        body: SignInWithEmailRequest
    ): AuthBaseResponse<UserCredentialResponse> {
        validationUtil.validate(body)

        val login = userCredentialRepository
            .findByUserEmail(
                body.email!!
            ) ?: throw UnAuthorizedException(translate("auth.user.not.exist"))

        if (!scrypt.matches(body.password, login.userPassword)) {
            throw UnAuthorizedException(translate("auth.invalid"))
        }

        if (login.userAuthProvider != BASIC.name) {
            throw UnAuthorizedException(translate("auth.method.not.allowed"))
        }

        val generatedToken = jwtUtil.generateToken(login.userEmail)


        return baseAuthResponse {
            code = OK.value()
            data = login.toResponse()
            message = translate("auth.success")
            token = generatedToken
        }

    }

    fun signInWithGoogle(
        request: SignInWithGoogleRequest
    ): AuthBaseResponse<UserCredentialResponse> {
        validationUtil.validate(request)
        val googleAuth = GoogleAuthUtil(environment, i18n)
        val verifyUser =
            googleAuth.getGoogleClaim(request.token!!)
        if (!verifyUser.valid) {
            throw UnAuthorizedException(verifyUser.message)
        }
        val findUser = userCredentialRepository.findByUserEmail(verifyUser.email)
            ?: throw UnAuthorizedException(translate("auth.user.not.exist"))

        if (findUser.userAuthProvider != UserAuthProvider.GOOGLE.name) {
            throw UnAuthorizedException(translate("auth.method.not.allowed"))
        }

        return baseAuthResponse {
            code = OK.value()
            data = findUser.toResponse()
            message = translate("auth.success")
        }
    }

    @Transactional
    fun signUpWithEmail(
        request: SignUpWithEmailRequest
    ): AuthBaseResponse<UserCredentialResponse> {
        validationUtil.validate(request)

        val isExist = userCredentialRepository.exist(request.email!!)
        if (isExist) {
            throw UnAuthorizedException(translate("auth.failed.user.exist"))
        }

        val uuid = UUID.randomUUID().toString()
        val date = getTodayDateTimeOffset()
        val activationToken = UUID.randomUUID().toString()
        val userProfile = UserProfile(
            userId = uuid,
            userFullName = request.fullName!!,
            userProfilePicture = null,
            userDateOfBirth = null,
            userCountryCode = "id",
            userPhoneNumber = null,
            createdAt = date,
            updatedAt = date
        )
        val savedProfile = userProfileRepository.save(userProfile)
        val userCredential = UserCredential(
            userId = uuid,
            userEmail = request.email!!,
            userPassword = scrypt.encode(request.password),
            userStatus = WAITING_CONFIRMATION.name,
            userAuthProvider = BASIC.name,
            userPermissions = listOf(),
            userProfile = savedProfile,
            userAuthTokenProvider = "",
            userNotificationToken = "",
            createdAt = date,
            updatedAt = date
        )
        val savedCredential = userCredentialRepository.save(userCredential)

        //create verification
        val verification = UserVerification(
            userActivationId = uuid,
            userActivationToken = activationToken,
            userId = uuid,
            createdAt = date,
            activationAt = null

        )
        val savedVerification = userVerificationRepository.save(verification)
        //todo send email


        return baseAuthResponse {
            code = OK.value()
            data = savedCredential.toResponse()
            message = translate("auth.success")
        }
    }

    @Transactional
    fun signUpWithGoogle(
        request: SignUpWithGoogleRequest
    ): AuthBaseResponse<UserCredentialResponse> {
        validationUtil.validate(request)

        val claim = GoogleAuthUtil(
            environment,
            i18n
        ).getGoogleClaim(request.token!!)

        if (!claim.valid) {
            throw UnAuthorizedException(claim.message)
        }

        val isExist = userCredentialRepository.exist(claim.email)
        if (isExist) {
            throw UnAuthorizedException(translate("auth.failed.user.exist"))
        }

        val uuid = UUID.randomUUID().toString()
        val date = getTodayDateTimeOffset()
        val userProfile = UserProfile(
            userId = uuid,
            userFullName = claim.fullName,
            userProfilePicture = null,
            userDateOfBirth = null,
            userCountryCode = "id",
            userPhoneNumber = null,
            createdAt = date,
            updatedAt = date
        )
        val savedProfile = userProfileRepository.save(userProfile)
        val userCredential = UserCredential(
            userId = uuid,
            userEmail = claim.email,
            userPassword = scrypt.encode(claim.email),
            userStatus = ACTIVE.name,
            userAuthProvider = BASIC.name,
            userPermissions = listOf(),
            userProfile = savedProfile,
            userAuthTokenProvider = request.token!!,
            userNotificationToken = "",
            createdAt = date,
            updatedAt = date
        )
        val savedCredential = userCredentialRepository.save(userCredential)

        return baseAuthResponse {
            code = OK.value()
            data = savedCredential.toResponse()
            message = translate("auth.success")
        }
    }

    @Transactional
    fun userVerification(
        token: String
    ): BaseResponse<List<Any>> {
        if (token.isEmpty()) {
            throw UnAuthorizedException("Verification failed")
        }

        val verificationData = userVerificationRepository.findByUserActivationToken(token)
            ?: throw UnAuthorizedException("Token not valid 1")

        if (verificationData.activationAt != null) {
            throw UnAuthorizedException("Token not valid 2")
        }

        if (Period.between(
                verificationData.createdAt.toLocalDate(),
                getTodayDateTime().toLocalDate()
            ).days > 1
        ) {
            throw UnAuthorizedException("Token not valid or expired")
        }

        val findUser = userCredentialRepository.findByIdOrNull(verificationData.userId)
            ?: throw UnAuthorizedException("User not found 3")

        userCredentialRepository.save(
            findUser.copy(
                userStatus = ACTIVE.name
            )
        )
        userVerificationRepository.save(
            verificationData.copy(
                activationAt = OffsetDateTime.now()
            )
        )

        return baseResponse {
            code = OK.value()
            data = listOf()
            message = translate("auth.verification.success")

        }
    }

    fun refreshToken(
        jwt: String?
    ): AuthBaseResponse<List<Any>> {
        if (jwt.isNullOrEmpty()) {
            throw UnAuthorizedException("Token not valid")
        }
        val claims = jwtUtil.decode(jwt)
        if (!claims.first) {
            throw UnAuthorizedException(claims.second)
        }

        val generate = jwtUtil.generateToken(claims.second)

        return baseAuthResponse {
            code = OK.value()
            data = listOf()
            message = translate("auth.success")
            token = generate
        }
    }

    fun uploadProfilePicture(
        request: UpdateProfilePictureRequest
    ): BaseResponse<UserProfileResponse> = buildResponse {
        val user = userCredentialRepository.findByUserEmail(it)
            ?: throw UnAuthorizedException(translate("auth.user.not.exist"))

        val mediaType = request.file?.getMimeTypes().orEmpty()
        val fileNAme = user.createFileName(mediaType)



        baseResponse { }
    }
    //end region

    fun getAllUsers(
        pageable: Pageable
    ): BaseResponse<PagingDataResponse<UserProfileResponse>> = buildResponse(
        allow = { it.allowTo(READ_USER) }
    ) {
        val findAll = userProfileRepository.findAll(pageable)

        baseResponse {
            code = OK.value()
            data = findAll.toResponse()
            message = translate("users.found")
        }
    }
}